### PR TITLE
NL Writer, version 2

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This PR introduces, but does not activate, a new NL writer.  The new (v2) writer is about 12% faster than the old (v1) writer.  Some key features of the new writer:
- Only makes a single pass through the model
- Only walk each expression once (build an intermediate string representation that supports deferring the naming of variables until after all constraints have been processed and the variables have been correctly categorized.
- Support emitting `Expression` components as AMPL "defined variables" (instead of expanding them into the constraint expression trees)
- Supports models where `Var` objects do not live on the block being solved
- Rigorously checks for and complains about unrecognized active Components

## Changes proposed in this PR:
- Add the `nl_v2` writer
- Rename `nl` writer to `nl_v1`, and make `nl` be an alias to the `nl_v1` writer.

## Other changes in this PR:
- Update the kernel constraint, conic, and matrix APIs to match Constraint (effectively reopening #2236)
- Update `StreamBasedExpressionVisitor` to be recursive by default, and fall back on the non-recursive version if the stack gets "too deep"
- Public methods to get / set direction and datatype for Suffix
- Fix (remove) incorrect implementation of `SumExpression.is_potentially_variable()` 
- Efficiency improvements setting / getting constraint and variable bounds
- Efficiency improvements getting bounds and intervals for global Sets
- Efficiency improvement in `SymbolMap.add_symbols()`
- Added a `GeneralTimer` object (complementary to ConstructionTimer and TransformationTimer)
- Update `tic` and `toc` to accept positional arguments to use to `%`-format the resulting message
- Update complementarity and mpec tests to write files to tmp (and not the source directory)
- Update the nl file determinism for existing tests for consistency between v1 and v2 writers
- Add a utility for improved comparisons of NL files

## Performance
This PR has two aspects of performance.  First, the rework of the `StreamBasedExpressionVisitor` produces a significant performance gain.  The following is the difference between `main` and this PR (the `nl` results are with the OLD v1 writer):

image

On top of that, the new v2 writer is about 12% faster than the old v1 writer:

image

Combined, merging this PR and activating the v2 writer speeds up the NL write time by ~19%:

image

(all above are comparing the minimum from 10 replicates)

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/base/constraint.py [hunk 1]
```diff
@@ -386,8 +390,10 @@ def upper(self):
     @property
     def lb(self):
         """Access the value of the lower bound of a constraint expression."""
-        bound = value(self._lb())
-        if bound is not None and not math.isfinite(bound):
+        bound = self._lb()
+        if bound.__class__ not in native_types:
+            bound = value(bound)
+        if bound in _nonfinite_values or bound != bound:
```
Comment (mrmundt) [inline pyomo/core/base/constraint.py:396]: When would `bound != bound` ever trigger?

File: pyomo/core/base/constraint.py [hunk 2]
```diff
@@ -399,8 +405,10 @@ def lb(self):
     @property
     def ub(self):
         """Access the value of the upper bound of a constraint expression."""
-        bound = value(self._ub())
-        if bound is not None and not math.isfinite(bound):
+        bound = self._ub()
+        if bound.__class__ not in native_types:
+            bound = value(bound)
+        if bound in _nonfinite_values or bound != bound:
```
Comment (mrmundt) [inline pyomo/core/base/constraint.py:411]: Same question as above.

File: pyomo/core/base/constraint.py [hunk 3]
```diff
@@ -614,21 +622,30 @@ def set_value(self, expr):
...
         # We have historically mapped incoming inf to None
         if self._lower.__class__ in native_numeric_types:
-            if self._lower == -_inf:
-                self._lower = None
-            elif not math.isfinite(self._lower):
-                raise ValueError(
-                    "Constraint '%s' created with an invalid non-finite "
-                    "lower bound (%s)." % (self.name, self._lower))
+            bound = self._lower
+            if bound in _nonfinite_values or bound != bound:
```
Comment (mrmundt) [inline pyomo/core/base/constraint.py:633]: Same question as above.

File: pyomo/core/base/constraint.py [hunk 4]
```diff
@@ -614,21 +622,30 @@ def set_value(self, expr):
...
+                        "lower bound (%s)." % (self.name, self._lower))
         if self._upper.__class__ in native_numeric_types:
-            if self._upper == _inf:
-                self._upper = None
-            elif not math.isfinite(self._upper):
-                raise ValueError(
-                    "Constraint '%s' created with an invalid non-finite "
-                    "upper bound (%s)." % (self.name, self._upper))
+            bound = self._upper
+            if bound in _nonfinite_values or bound != bound:
```
Comment (mrmundt) [inline pyomo/core/base/constraint.py:642]: Same question as above.

File: pyomo/core/expr/visitor.py [hunk 5]
```diff
@@ -183,8 +221,260 @@ def wrapper(*args):
...
+        # will mock up the "with args" using a try-finally.
+        context_manager = hasattr(args, '__enter__')
+        if context_manager:
+            args.__enter__()
+
+        try:
+            descend = True
+            child_idx = -1
+            # Note: this relies on iter(iterator) returning the
+            # iterator.  This seems to hold got all common iterators
```
Comment (mrmundt) [inline pyomo/core/expr/visitor.py:313]: Typo: `got` -> `for`

File: pyomo/core/expr/visitor.py [hunk 6]
```diff
@@ -183,8 +221,260 @@ def wrapper(*args):
...
+        # Because we do not require the args to be a context manager, we
+        # will mock up the "with args" using a try-finally.
+        context_manager = hasattr(args, '__enter__')
+        if context_manager:
+            args.__enter__()
+
+        try:
+            child_idx = -1
+            # Note: this relies on iter(iterator) returning the
+            # iterator.  This seems to hold got all common iterators
```
Comment (mrmundt) [inline pyomo/core/expr/visitor.py:385]: Typo: `got` -> `for`

File: pyomo/core/expr/visitor.py [hunk 7]
```diff
@@ -183,8 +221,260 @@ def wrapper(*args):
...
+
+        # We are done with this node.  Call exitNode to compute
+        # any result
+        if self.exitNode is not None:
+            return self.exitNode(node, data)
+        else:
+            return data
+
+    def _process_node_bex(self, node, recursion_limit):
+        """Recursive routine for processing nodes with only bex callbacks
```
Comment (mrmundt) [inline pyomo/core/expr/visitor.py:352]: May want to define `bex` and `bx`

File: pyomo/core/kernel/conic.py [hunk 8]
```diff
@@ -97,14 +97,24 @@ def body(self):
...
+
+    @property
+    def upper(self):
+        """The expression for the upper bound of the constraint"""
+        return 0.0
+
     @property
     def lb(self):
-        """The lower bound of the constraint"""
+        """The value of the lower lower bound of the constraint"""
```
Comment (mrmundt) [inline pyomo/core/kernel/conic.py:112]: Of the lower lower bound?

File: pyomo/core/kernel/conic.py [hunk 9]
```diff
@@ -97,14 +97,24 @@ def body(self):
...
         return None
 
     @property
     def ub(self):
-        """The upper bound of the constraint"""
+        """The value of the lower upper bound of the constraint"""
```
Comment (mrmundt) [inline pyomo/core/kernel/conic.py:117]: Of the lower upper bound?

File: pyomo/core/tests/unit/test_visitor.py [hunk 10]
```diff
@@ -13,14 +13,20 @@
 #
 
 import os
-import pyomo.core.tests.unit.test_visitor
-from os.path import abspath, dirname
-currdir = dirname(abspath(__file__))+os.sep
+import platform
+import sys
+
+#import pyomo.core.tests.unit.test_visitor
```
Comment (mrmundt) [inline pyomo/core/tests/unit/test_visitor.py:19]: Residual comment?

File: pyomo/repn/plugins/ampl/ampl_.py [hunk 11]
```diff
@@ -360,6 +361,10 @@ def __call__(self,
         export_nonlinear_variables = \
             io_options.pop("export_nonlinear_variables", False)
 
+        # column_order is a new option supported by the nl wraiter v2
```
Comment (mrmundt) [inline pyomo/repn/plugins/ampl/ampl_.py:364]: Typo: `writer`

File: pyomo/repn/plugins/nl_writer.py [hunk 12]
```diff
@@ -0,0 +1,2124 @@
...
+            # matches the file_determinism flag.  This is a little
+            # cumbersome, but is implemented this way for consistency
+            # with the original NL writer.
+            if self.config.column_order is None:
+                self.config.column_order = []
+            self.config.column_order.extend(model.component_data_objects(
+                Var, descend_into=True, sort=sorter))
+        if self.config.column_order is not None:
+            # Note that Vars that appear twice (e.g., through a
+            # Reference) will be sorted with the FIRST occurance.
```
Comment (mrmundt) [inline pyomo/repn/plugins/nl_writer.py:397]: Typo: `occurrence`

File: pyomo/repn/plugins/nl_writer.py [hunk 13]
```diff
@@ -0,0 +1,2124 @@
...
+        ostream.write(''.join(_init_lines))
+
+        #
+        # "r" lines (constraint bounds)
+        #
+        ostream.write('r%s\n' % (
+            "\t#%d ranges (rhs's)" % len(constraints)
+            if symbolic_solver_labels else '',
+        ))
+        # _bound_writer = {
```
Comment (mrmundt) [inline pyomo/repn/plugins/nl_writer.py:1039]: Intentionally left here?

File: pyomo/repn/plugins/nl_writer.py [hunk 14]
```diff
@@ -0,0 +1,2124 @@
...
+        all_linear_vars = set()
+        all_nonlinear_vars = set()
+        nnz_by_var = {}
+
+        for comp_info in comp_list:
+            expr_info = comp_info[1]
+            # Note: mult will be 1 here: it is either cleared by
+            # finalizeResult, or this is a named expression, in which
+            # case the mult was reset within handle_named_expression_node
+            #assert expr_info.mult == 1
```
Comment (mrmundt) [inline pyomo/repn/plugins/nl_writer.py:1176]: For debugging?

File: pyomo/repn/plugins/nl_writer.py [hunk 15]
```diff
@@ -0,0 +1,2124 @@
...
+                    linear = {}
+                    for v, c in expr_info.linear:
+                        if v in linear:
+                            linear[v] += c
+                        else:
+                            linear[v] = c
+                    expr_info.linear = linear
+                linear_vars = set(expr_info.linear)
+                all_linear_vars.update(linear_vars)
+            # else:
```
Comment (mrmundt) [inline pyomo/repn/plugins/nl_writer.py:1190]: Still necessary?

File: pyomo/repn/plugins/nl_writer.py [hunk 16]
```diff
@@ -0,0 +1,2124 @@
...
+                    for i in nonlinear_vars - linear_vars:
+                        expr_info.linear[i] = 0
+                else:
+                    # All variables are nonlinear; generate the linear
+                    # dict with all zeros
+                    expr_info.linear = dict.fromkeys(nonlinear_vars, 0)
+                all_nonlinear_vars.update(nonlinear_vars)
+
+            # Update the count of components that each variable appears in
+            #nnz_by_var.update(nz)
```
Comment (mrmundt) [inline pyomo/repn/plugins/nl_writer.py:1218]: Debugging?

File: pyomo/repn/plugins/nl_writer.py [hunk 17]
```diff
@@ -0,0 +1,2124 @@
...
+            elif info[1] is None:
+                n_subexpressions[3 if info[0] else 1] += 1
+            elif info[0] is None:
+                n_subexpressions[4 if info[1] else 2] += 1
+            else:
+                n_subexpressions[0] += 1
+        return n_subexpressions
+
+    def _write_nl_expression(self, repn, include_const):
+        #assert repn.mult == 1
```
Comment (mrmundt) [inline pyomo/repn/plugins/nl_writer.py:1301]: Debugging?

File: pyomo/repn/plugins/nl_writer.py [hunk 18]
```diff
@@ -0,0 +1,2124 @@
...
+        This method assumes that the operator was "+". It is implemented
+        so that we can directly use an AMPLRepn() as a data object in
+        the expression walker (thereby avoiding the function call for a
+        custom callback)
+
+        """
+        # Note that self.mult will always be 1 (we only call apend()
+        # within a sum, so there is no opportunity for self.mult to
+        # change)
+        #assert self.mult == 1
```
Comment (mrmundt) [inline pyomo/repn/plugins/nl_writer.py:1452]: Debugging?

File: pyomo/repn/plugins/nl_writer.py [hunk 19]
```diff
@@ -0,0 +1,2124 @@
...
+        visitor, visitor.template.and_expr + op[0])
+    nl2, args2 = node_result_to_amplrepn(arg2).compile_repn(visitor)
+    nl += nl2 + op[1] + nl2
+    args.extend(args2)
+    args.extend(args2)
+    nonlin = node_result_to_amplrepn(arg3).compile_repn(visitor, nl, args)
+    return (_GENERAL, AMPLRepn(0, None, nonlin))
+
+def handle_named_expression_node(visitor, node, arg1):
+    #return arg1
```
Comment (mrmundt) [inline pyomo/repn/plugins/nl_writer.py:1692]: Debugging?

File: pyomo/repn/plugins/nl_writer.py [hunk 20]
```diff
@@ -0,0 +1,2124 @@
...
+def _before_var(visitor, child):
+    _id = id(child)
+    if _id not in visitor.var_map:
+        if child.fixed:
+            return False, (_CONSTANT, child())
+        visitor.var_map[_id] = child
+    return False, (_MONOMIAL, _id, 1)
+
+def _before_npv(visitor, child):
+    # _id = id(child)
```
Comment (mrmundt) [inline pyomo/repn/plugins/nl_writer.py:1888]: Debugging?

File: pyomo/repn/plugins/nl_writer.py [hunk 21]
```diff
@@ -0,0 +1,2124 @@
...
+    return False, (_CONSTANT, child())
+
+def _before_monomial(visitor, child):
+    #
+    # The following are performance optimizations for common
+    # situations (Monomial terms and Linear expressions)
+    #
+    arg1, arg2 = child._args_
+    if arg1.__class__ not in native_types:
+        # _id = id(arg1)
```
Comment (mrmundt) [inline pyomo/repn/plugins/nl_writer.py:1902]: Debugging?

**Review by mrmundt — CHANGES_REQUESTED** [review]

File: pyomo/core/base/constraint.py (hunk 1 above)
Comment (jsiirola) [reply pyomo/core/base/constraint.py:396]: If bound is `float('nan')`, then `(bound != bound) is True`

File: pyomo/core/base/constraint.py (hunk 2 above)
Comment (jsiirola) [reply pyomo/core/base/constraint.py:411]: If bound is `float('nan')`, then `(bound != bound) is True`

File: pyomo/core/base/constraint.py (hunk 3 above)
Comment (jsiirola) [reply pyomo/core/base/constraint.py:633]: If bound is `float('nan')`, then `(bound != bound) is True`

File: pyomo/core/base/constraint.py (hunk 4 above)
Comment (jsiirola) [reply pyomo/core/base/constraint.py:642]: If bound is `float('nan')`, then `(bound != bound) is True`

File: pyomo/repn/plugins/nl_writer.py (hunk 15 above)
Comment (jsiirola) [reply pyomo/repn/plugins/nl_writer.py:1190]: Yes.  This is actually documentation as to why the (seemingly necessary symbol definition is not actually necessary).  Updating comment to clarify.

File: pyomo/core/base/constraint.py (hunk 1 above)
Comment (jsiirola) [reply pyomo/core/base/constraint.py:396]: (this and the other three instances of this idiom are covered by tests in `pyomo/core/tests/unit/test_con.py`)

File: pyomo/common/timing.py [hunk 22]
```diff
@@ -198,22 +218,33 @@ def tic(self, msg=_NotSpecified,
...
             msg = "Resetting the tic/toc delta timer"
         if msg is not None:
-            self.toc(msg=msg, delta=False, ostream=ostream, logger=logger)
-
-
-    def toc(self, msg=_NotSpecified, delta=True,
-            ostream=_NotSpecified, logger=_NotSpecified):
+            if args and '%' not in msg:
+                deprecation_warning(
+                    "'ostream', and 'logger' should be specified "
```
Comment (mrmundt) [inline pyomo/common/timing.py:237]: Small peeve... I don't like the comma between `'ostream', and 'logger'`. Don't need to fix but it annoys me.

**Review by mrmundt — APPROVED** [review]

File: pyomo/core/expr/visitor.py [hunk 23]
```diff
@@ -885,13 +1190,20 @@ def evaluate_expression(exp, exception=True, constant=False):
...
         visitor = _EvaluateConstantExpressionVisitor()
     else:
-        visitor = _EvaluationVisitor(exception=exception)
+        if evaluate_expression.visitor_active:
+            visitor = _EvaluationVisitor(exception=exception)
+        else:
+            visitor = evaluate_expression.visitor_cache
+            visitor.exception = exception
+            evaluate_expression.visitor_active = True
+            clear_active = True
```
Comment (michaelbynum) [inline pyomo/core/expr/visitor.py:1203]: What is the purpose of these changes? Avoid reconstruction of the visitor each time the function is called?

File: pyomo/repn/plugins/nl_writer.py [hunk 24]
```diff
@@ -0,0 +1,2128 @@
...
+            SORT_INDICES (2) : sort keys of indexed components
+            SORT_SYMBOLS (3) : sort keys AND sort names (over declaration order)
+        """
+    ))
+    CONFIG.declare('symbolic_solver_labels', ConfigValue(
+        default=False,
+        domain=bool,
+        description='Write the corresponding .row and .col files',
+    ))
+    CONFIG.declare('export_nonlinear_variables', ConfigValue(
```
Comment (michaelbynum) [inline pyomo/repn/plugins/nl_writer.py:157]: Should this be renamed? Maybe something like `export_unused_variables`?

File: pyomo/repn/plugins/nl_writer.py [hunk 25]
```diff
@@ -0,0 +1,2128 @@
...
+        which appear before variables appearing only in objectives.
+        Within each group, continuous variables appear before discrete
+        variables.  In all cases, column_order is preserved within each
+        group."""
+    ))
+
+    def __init__(self):
+        self.config = self.CONFIG()
+
+    def __call__(self, model, filename, solver_capability, io_options):
```
Comment (michaelbynum) [inline pyomo/repn/plugins/nl_writer.py:192]: Please document the arguments and return value for this method.

Comment (michaelbynum) [reply pyomo/repn/plugins/nl_writer.py:192]: Can this method please also return the ordering of the variables and the constraints (or instructions on how to access this.

File: pyomo/repn/tests/ampl/nl_diff.py [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+    # Normalize consecutive whitespace to a single space
+    data = _norm_whitespace.sub(' ', data)
+    # preface all comments with a single tab character
+    data = _norm_comment.sub('\t#', data)
+    # Normalize floating point integers to integers
+    data = _norm_integers.sub('', data)
+    # return the sequence of lines
+    return data.splitlines()
+
+def nl_diff(base, test, baseline='baseline', testfile='testfile'):
```
Comment (michaelbynum) [inline pyomo/repn/tests/ampl/nl_diff.py:81]: Very nice!

File: pyomo/core/expr/numeric_expr.py [hunk 27]
```diff
@@ -1053,14 +1053,6 @@ def is_constant(self):
...
         return False
 
-    def is_potentially_variable(self):
-        for v in islice(self._args_, self._nargs):
-            if v.__class__ in nonpyomo_leaf_types:
-                continue
-            if v.is_variable_type() or v.is_potentially_variable():
-                return True
-        return False
-
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/core/expr/numeric_expr.py:1063]: We rely on `is_potentially_variable()` much like `is_constant()`: `False` needs to be a promise that the node can *never* be potentially variable (this comes into play with generating things like `MonomialTermExpression`: if `is_potentially_variable() is false, the term is allowed to be in the "coef" side of a monomial term.

File: pyomo/repn/plugins/nl_writer.py (hunk 24 above)
Comment (jsiirola) [reply pyomo/repn/plugins/nl_writer.py:157]: I would love to change this name.  It is here for backwards compatibility.

File: pyomo/core/expr/visitor.py (hunk 23 above)
Comment (jsiirola) [reply pyomo/core/expr/visitor.py:1203]: Yes - but still make this thread / recursion-safe.

Comment (michaelbynum) [reply pyomo/core/expr/visitor.py:1203]: Ah. Got it. Thanks.

File: pyomo/repn/plugins/nl_writer.py (hunk 24 above)
Comment (michaelbynum) [reply pyomo/repn/plugins/nl_writer.py:157]: Oh. I've never used that option. I didn't realize it existed. Another reason to add an alias feature to the config system...

File: pyomo/repn/plugins/nl_writer.py (hunk 25 above)
Comment (jsiirola) [reply pyomo/repn/plugins/nl_writer.py:192]: I will add some documentation as soon as I can figure out what the arguments are.  ;)  This API can't really be changed: it is here for backwards compatibility with the old writers.  The more I work with core, the more I would like to deprecate most uses of `__call__()`.

File: pyomo/repn/plugins/nl_writer.py (hunk 24 above)
Comment (jsiirola) [reply pyomo/repn/plugins/nl_writer.py:157]: Agreed.  FWIW, I think this was added for PyNumero and ExternalGreyBoxModel: you have variables that only appear in external functions, so you need to "force" the NL writer to include variables that otherwise look fixed to the writer.

File: pyomo/repn/plugins/nl_writer.py (hunk 25 above)
Comment (michaelbynum) [reply pyomo/repn/plugins/nl_writer.py:192]: Haha!

Can we create a public API and then wrap it for use with the `WriterFactory` and backwards compatibility (similar to what we are doing for Appsi)?

**Comment by jsiirola** [comment]
Sure - but at the moment, I think the public API would just be to remove `__call__`.  It may just be easier to update Pyomo to not use call and then just deprecate the method (all in future PRs)

File: pyomo/repn/plugins/nl_writer.py (hunk 25 above)
Comment (michaelbynum) [reply pyomo/repn/plugins/nl_writer.py:192]: Oh! My bad. I see. I just want the `write` method, and the variable constraint ordering is in `NLWriterInfo`? Maybe I really just want to you document the `write` method and `NLWriterInfo`. I'm not sure I care about documenting `__call__`.

File: pyomo/repn/plugins/nl_writer.py [hunk 28]
```diff
@@ -0,0 +1,2128 @@
...
+            row_labels, col_labels)
+        timer.toc("Wrote NL stream", level=logging.DEBUG)
+        timer.toc("Generated NL representation", delta=False)
+        return info
+
+    def _categorize_vars(self, comp_list, linear_by_comp):
+        """Categorize compiled expression vars into linear and nonlinear
+
+        This routine takes an iterable of compiled component expression
+        infos and returns the sets of variables appearing lineary and
```
Comment (blnicho) [inline pyomo/repn/plugins/nl_writer.py:1130]: Typo: linearly

File: pyomo/repn/plugins/nl_writer.py [hunk 29]
```diff
@@ -0,0 +1,2128 @@
...
+    def _write_v_line(self, expr_id, k):
+        ostream = self.ostream
+        column_order = self.column_order
+        info = self.subexpression_cache[expr_id]
+        if self.symbolic_solver_labels:
+            lbl = '\t#%s' % info[0].name
+        else:
+            lbl = ''
+        self.var_id_to_nl[expr_id] = f"{self.next_V_line_id}{lbl}"
+        # Do NOT write out 0 coeffficients here: doing so fouls up the
```
Comment (blnicho) [inline pyomo/repn/plugins/nl_writer.py:1322]: Typo: coefficients

File: pyomo/repn/plugins/nl_writer.py [hunk 30]
```diff
@@ -0,0 +1,2128 @@
...
+
+        Notes
+        -----
+        This method assumes that the operator was "+". It is implemented
+        so that we can directly use an AMPLRepn() as a data object in
+        the expression walker (thereby avoiding the function call for a
+        custom callback)
+
+        """
+        # Note that self.mult will always be 1 (we only call apend()
```
Comment (blnicho) [inline pyomo/repn/plugins/nl_writer.py:1445]: Typo: append()

File: pyomo/repn/plugins/nl_writer.py [hunk 31]
```diff
@@ -0,0 +1,2128 @@
...
+        if component_map[Suffix]:
+            if not row_order:
+                row_order = {id(con[0]): i for i, con in enumerate(constraints)}
+            obj_order = {id(obj[0]): i for i, obj in enumerate(objectives)}
+            model_id = id(model)
+            # Note: reverse the block list so that higher-level Suffix
+            # components override lower level ones.
+            for block in reversed(component_map[Suffix]):
+                for suffix in block.component_objects(
+                        Suffix, active=True, descend_into=False, sort=sorter):
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/repn/plugins/nl_writer.py:734]: This won't resolve #672 in general, but it should for Ipopt (and any other NL solver)

File: pyomo/repn/plugins/nl_writer.py [hunk 32]
```diff
@@ -0,0 +1,2153 @@
...
+            Could be an opened file or a io.StringIO.
+
+        rowstream: io.TextIOBase
+            A text output stream to write the ASL "row file" (list of
+            constraint / objective names).  Ignored unless
+            `symbolic_solver_labels` is True.
+
+        colstream: io.TextIOBase
+            A text output stream to write the ASL "col file" (list of
+            variable names).  Ignored unless `symbolic_solver_labels` is True.
```
Comment (michaelbynum) [inline pyomo/repn/plugins/nl_writer.py:241]: Thanks for this. What about the return type?

**Review by michaelbynum — APPROVED** [review]
